YAML refactor to support gRPC

glennmatthews opened the pull request:
Note that this is a pull request to merge code into feature/grpc - not yet to develop.

Substantial refactor of the YAML files for CommandReference. Key features:
- YAML files are autodiscovered from the `cmd_ref/` subdirectory
- Each YAML file describes a feature (and all of its platform variants) instead of each YAML file describing all features for a single platform - more maintainable.
- Token generation (`_append`, printf wildcards, key-value wildcards) is now part of CommandReference instead of being part of Node.
- New YAML constructs to handle API variants (`nxapi:`, `grpc:`) and platform variants (`/N7K/:`, `/N9K/:`, `else:`)

NXAPI minitests are passing except for test_node.rb (which needs some massaging still).

Commit:
Class variable to constant

Commit:
YAML refactor, part one

Instead of grouping YAML by platform, group it by feature.
(We've already started this somewhat in feature/bgp branch)
Allow per-API and per-platform overrides within the YAML.
test_node_ext now *mostly* passes for gRPC.
More fixes yet to come.

Commit:
Move wildcard logic from Node to CommandReference

test_node_ext still passes on NXOS.
test_router_ospf (a major user of templating logic) also passes.

Commit:
Get test_command_reference passing

Commit:
Test_command_reference code coverage

Commit:
Add '?' APIs to CommandReference

These let you check whether a given element is set at all,
for example 'ref.config_get?'.
Fix errors in interface/vlan test by using config_get? where needed.

Commit:
Documentation fixes

Commit:
More documentation fixes

Commit:
Different default intfs for nxapi/grpc

Commit:
Use template more

mikewiebe commented on docs/README-develop-best-practices.md:
````diff
@@ -43,132 +44,124 @@ This document is intended to assist in developing cisco_node_utils API's that ar
 ```
 
-### <a name="yaml3">Y3: Avoid nested optional matches.
+### <a name="yaml4">Y4: Avoid nested optional matches.
````
Since you are already modifying this file, could you add an example for this?  I meant to, but missed it.

glennmatthews replied:
Added.

robert-w-gries commented:
Travis is currently failing to build due to the nxapi gem name change.  Can you manually run rubocop and post the results?

glennmatthews commented:
Rubocop is passing in full - the git hooks run it automatically for every commit I do.

Commit:
Add example for Y4 rule

chrisvanheuveln commented on docs/README-develop-best-practices.md:
```diff
@@ -43,132 +44,145 @@ This document is intended to assist in developing cisco_node_utils API's that ar
-### <a name="yaml7">Y7: Use Key-value wildcards instead of Printf-style wildcards.
+### <a name="yaml8">Y8: Use Key-value wildcards instead of Printf-style wildcards.
 
 The following approach is moderately more complex to implement but is more readable in the ruby code and is flexible enough to handle significant platform differences in CLI. It is therefore the recommended approach for new development.
```
Consider:
"Key-value wildcards are moderately more complex to implement than Printf-style wildcards but they are more readable in the ruby code and are flexible enough to handle significant platform differences in CLI. Key-value wildcards are therefore the recommended approach for new development."

chrisvanheuveln commented on docs/README-develop-best-practices.md:
```diff
@@ -239,7 +253,7 @@ Having this logic defined in the common object lets the minitest easily check th
 
 Without this equality operator `==` only passes if they are the same instance object.  With this equality operator `==` passes if they are different objects referring to the same configuration on the node.
```
"The built-in equality operator `==` returns true only if they are the same instance object.  The `==`method below is used to override the built-in equality operator, which returns true even when they are different objects referring to the same configuration on the node."

mikewiebe commented on lib/cisco_node_utils/cmd_ref/README_YAML.md:
```diff
@@ -0,0 +1,458 @@
+filename:
+  config_get: "show running vtp"
+  config_get_token: '/vtp file (\S+)/'
+  config_set: "%s vtp password %s"
```
Given that we are recommending wildcard substitution, I would prefer we use it in this and other examples.  Except the example showing Printf-style wildcards of course :)

mikewiebe commented on lib/cisco_node_utils/cmd_ref/README_YAML.md:
```diff
@@ -0,0 +1,458 @@
+
+### API variants
+
+Any of the attribute properties can be subdivided by client API using the
```
I feel like this needs more context.  Can we add a reference pointer that describes the client API options in more detail?

glennmatthews replied:
Will flesh this out a bit more in the next diff - not sure what kind of reference pointer you are looking for but hopefully the extra detail renders that moot.

mikewiebe commented on lib/cisco_node_utils/cmd_ref/README_YAML.md:
````diff
@@ -0,0 +1,458 @@
+for all products that do not match any of the given regexps:
+
+```yaml
+# show_version.yaml
````
I like this structure much better then our old structure!  Nice improvement.

chrisvanheuveln replied:
Nice work. Aside from the nits it's good to go.

robert-w-gries commented on tests/test_command_reference.rb:
```diff
@@ -42,44 +43,40 @@ def write(string)
 
+  def test_data_sanity
+    # Make sure the actual YAML in our library loads for various platforms
+    CommandReference.new('', '')
```
What is this testcase testing, and what are the expected results?

glennmatthews replied:
This is testing that we can create a CommandReference instance (using the 'live' YAML files) that doesn't refer to any specific API or platform, without any errors being thrown.
The following lines are similarly creating CommandReference instances for various known APIs and platforms.

Basically, it's a sanity check for our YAML files, as if there's any syntax error in any of them, this should catch it.

glennmatthews commented:
New diffs up to address Mike's and Chris's comments - thanks!

Commit:
Review feedback

- Reword several paragraphs for readability
- Use key-value wildcards in all examples (except printf-specific)
- More detail about API variants

glennmatthews commented:
Just wanted to note - since we aren't supporting Ruby < 2.0.0 any more, we are now guaranteed to have the new, improved 'Psych' YAML library instead of the older 'Syck' library that older releases used. This let me cut out some of the alternate code paths to handle Syck, and let me re-enable the logic that lets us check whether the YAML entries are alphabetized and complain if they aren't.

mikewiebe commented on lib/cisco_node_utils/cmd_ref/README_YAML.md:
```diff
@@ -0,0 +1,457 @@
+
+### API variants
+
+Clients for different Cisco platforms may use different APIs. Currently these include NXAPI (used for Cisco Nexus platforms) and gRPC (used for Cisco platforms running IOS XR). Often the CLI or other input/output formats needed will vary between APIs, so the YAML must be able to accomodate this.
```
Much better.

mikewiebe commented on lib/cisco_node_utils/cmd_ref/README_YAML.md:
```diff
@@ -0,0 +1,457 @@
+or an array of regexs.
+
+If this value is a string or array of strings, then the `config_get` command
+will be executed to produce _structured_ output and the string(s) will be
```
We should probably add a word of caution to make sure that _structured_ output works across all target platforms.

glennmatthews replied:
Added:

**WARNING: structured output, although elegant, may not be supported for all commands or all platforms. Use with caution.**

mikewiebe commented on lib/cisco_node_utils/cmd_ref/README_YAML.md:
```diff
@@ -0,0 +1,457 @@
+
+### `default_value`
+
+If there is a default value for this attribute when not otherwise specified by
```
More of a question really, but do we need to add guidance about when to use ~ as a default?

glennmatthews replied:
Yeah seems like a good suggestion - got any verbage to suggest? :)

mikewiebe replied:
Maybe....

For properties that have no default value set the default value to `default_value: ~` which indicates NULL (no value) in YAML.

glennmatthews replied:
Rewrote this section:

### `default_value`

If there is a default value for this attribute when not otherwise specified by
the user, the `default_value` parameter describes it. This can be a string,
boolean, integer, array, or nil.

``` yaml
description:
  default_value: ''

hello_interval:
  default_value: 10

auto_cost:
  default_value: [40, 'Gbps']

ipv4_address:
  # YAML represents nil as ~
  default_value: ~
```

By convention, a `default_value` of `''` (empty string) represents a configurable property that defaults to absent, while a default of `nil` (Ruby) or `~` (YAML) represents a property that has no meaningful default at all.

mikewiebe commented on lib/cisco_node_utils/cmd_ref/interface.yaml:
```diff
@@ -0,0 +1,236 @@
+    config_set_append: "%s negotiate auto"
+    default_value: true
+  test_config_get_regex: [
+  !ruby/regexp '/^\s+no negotiate auto/',
```
Is the !ruby/regexp still required or can we remove those?

glennmatthews replied:
It was still required for test_config_get_regex (only) but I've fixed that - new diffs incoming.

mikewiebe commented on lib/cisco_node_utils/command_reference.rb:
```diff
@@ -150,7 +180,53 @@ def self.debug=(value)
-    attr_reader :files, :product_id
+    KNOWN_APIS = %w(nxapi grpc)
+
+    def self.hash_merge(input_hash, api, product_id, base_hash=nil)
```
I think this method could use some documentation.  Lot happening here that is not completely obvious.

glennmatthews replied:
Done. New diffs are up.

Commit:
More review feedback

Document default_value nil vs '' usage
Warning about using structured output
Comment new CommandReference methods
Automatically convert test_config_get_regex values to Regexp

chrisvanheuveln commented:
:+1:

mikewiebe commented:
:+1:
